src/components/draggable/draggable-container.component.tsx

```diff
@@ -0,0 +1,155 @@
+   *
+   * `<DraggableItem />` is required to make `Draggable` works
+   */
+  children?: React.ReactNode;
+}
+
+interface DropItem {
+  id: number | string;
+  originalIndex: number;
+}
+
+const DropTarget = ({
+  children,
+  getOrder,
+  ...rest
+}: { getOrder: (movedItemId?: string | number | undefined) => void } & Omit<
+  DraggableContainerProps,
+  "getOrder"
+>) => {
+  const [, drop] = useDrop({
+    accept: "draggableItem",
+    drop(item: DropItem) {
+      // istanbul ignore else
+      if (getOrder) {
+        getOrder(item?.id);
+      }
+    },
+  });
+
+  return (
+    <StyledDraggableContainer ref={drop} {...rest}>
+      {children}
+    </StyledDraggableContainer>
+  );
+};
+
+const DraggableContainer = ({
+  children,
+  getOrder,
+  ...rest
+}: DraggableContainerProps): JSX.Element => {
+  const [draggableItems, setDraggableItems] = useState(
+    React.Children.toArray(children)
+  );
+
+  const hasProperChildren = useMemo(() => {
+    let hasDraggableItemChildren = true;
+
+    React.Children.toArray(children).forEach((child) => {
+      if (
+        typeof child === "string" ||
+        (React.isValidElement(child) &&
+          (child.type as React.FunctionComponent).displayName !==
+            DraggableItem.displayName)
+      ) {
+        hasDraggableItemChildren = false;
+      }
+    });
+
+    return hasDraggableItemChildren;
+  }, [children]);
+
+  // `<DraggableItem />` is required to make `Draggable` work
+  invariant(
+    hasProperChildren,
+    `\`${DraggableContainer.displayName}\` only accepts children of type \`${DraggableItem.displayName}\`.`
+  );
+
+  const isFirstRender = useRef(true);
+
+  useEffect(() => {
+    if (!isFirstRender.current) {
+      setDraggableItems(React.Children.toArray(children));
+    } else {
+      isFirstRender.current = false;
+    }
+  }, [children]);
+
+  const findItem = (id: string | number) => {
+    const draggableItem = draggableItems.filter((item: any) => {
+      return `${item.props.id}` === id;
+    })[0];
+
+    return {
+      draggableItem,
+      index: draggableItems.indexOf(draggableItem),
+    };
+  };
+
+  const moveItem = (id: string, atIndex: number) => {
+    const { draggableItem, index } = findItem(id);
+    if (!draggableItem) return;
+
+    const copyOfDraggableItems = [...draggableItems];
+    copyOfDraggableItems.splice(index, 1);
+    copyOfDraggableItems.splice(atIndex, 0, draggableItem);
+    setDraggableItems(copyOfDraggableItems);
+  };
+
+  const getItemsId = (item?: string | number) => {
+    if (!getOrder) {
+      return;
+    }
+
+    const draggableItemIds = draggableItems.map(
+      (draggableItem) => (draggableItem as { props: { id: string } }).props.id
+    );
+
+    getOrder(draggableItemIds, item);
+  };
+
+  const marginProps = filterStyledSystemMarginProps(rest);
+
+  return (
+    <DndProvider backend={HTML5Backend}>
+      <DropTarget getOrder={getItemsId} {...marginProps}>
+        {draggableItems.map((item: any) => {
```

Comment:
I'm not happy with this `any` typing here. The only other away around this I could think of was to use `React.isValidElement()` as this allows the use of `.props.`. The issue with this was that I'd need a way of returning a value for if the element is valid and if it isn't. By the nature of this component, it only allows `DraggableItems` to be children. So if I try to pass in a child that is a string or number, the invariant fires before it will ever reach this code. Open to suggestions.

Reply:
This should be fine to use unless I've missed something

```
{draggableItems.map((item) => {
          return React.isValidElement(item) && React.cloneElement(
            item as React.ReactElement,
            {
              id: `${item.props.id}`,
              findItem,
              moveItem,
            },
            [item.props.children]
          );
        })}
```